src/main/java/gregtech/common/metatileentities/converter/ConverterTrait.java

```diff
@@ -0,0 +1,350 @@
+            storedEU += voltage * maxAmps;
+            usedAmps += maxAmps;
+            return maxAmps;
+        }
+
+        @Override
+        public boolean inputsEnergy(EnumFacing side) {
+            return !feToEu && side != metaTileEntity.getFrontFacing();
+        }
+
+        @Override
+        public long changeEnergy(long amount) {
+            if (amount == 0)
+                return 0;
+            return amount > 0 ? addEnergy(amount) : removeEnergy(-amount);
+        }
+
+        @Override
+        public long addEnergy(long energyToAdd) {
+            if (energyToAdd <= 0)
+                return 0;
+            long original = energyToAdd;
+
+            // add energy first to internal buffer
+            long change = Math.min(baseCapacity - storedEU, energyToAdd);
+            storedEU += change;
+            energyToAdd -= change;
+
+            // then to batteries
+            energyToAdd -= chargeBattery(energyToAdd, false);
+
+            return original - energyToAdd;
+        }
+
+        @Override
+        public long removeEnergy(long energyToRemove) {
+            return extractInternal(energyToRemove, false);
+        }
+
+        @Override
+        public long getEnergyStored() {
+            long energyStored = storedEU;
+            IItemHandlerModifiable inventory = getInventory();
+            for (int i = 0; i < inventory.getSlots(); i++) {
+                IElectricItem electricItem = getBatteryContainer(inventory.getStackInSlot(i));
+                if (electricItem == null) continue;
+                energyStored = LongMath.saturatedAdd(electricItem.getCharge(), energyStored);
+                if (energyStored == Long.MAX_VALUE) break;
+            }
+            return energyStored;
+        }
+
+        @Override
+        public long getEnergyCapacity() {
+            long energyStored = baseCapacity;
+            IItemHandlerModifiable inventory = getInventory();
+            for (int i = 0; i < inventory.getSlots(); i++) {
+                IElectricItem electricItem = getBatteryContainer(inventory.getStackInSlot(i));
+                if (electricItem == null) continue;
+                energyStored = LongMath.saturatedAdd(electricItem.getMaxCharge(), energyStored);
+                if (energyStored == Long.MAX_VALUE) break;
+            }
+            return energyStored;
+        }
+
+        @Override
+        public long getInputAmperage() {
+            int inputAmperage = 0;
+            IItemHandlerModifiable inventory = getInventory();
+            for (int i = 0; i < inventory.getSlots(); i++) {
+                ItemStack batteryStack = inventory.getStackInSlot(i);
+                IElectricItem electricItem = getBatteryContainer(batteryStack);
+                if (electricItem == null) continue;
+                inputAmperage++;
+            }
+            return inputAmperage > 0 ? inputAmperage : baseAmps;
+        }
```

Comment:
I think base amps should always be 1 and total amperage should be 1 base amp + amount of batteries (* 2 (?))